addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/discovery/HomematicDeviceDiscoveryService.java

```diff
@@ -0,0 +1,175 @@
+ *
+ * @author Gerhard Riegler - Initial contribution
+ */
+public class HomematicDeviceDiscoveryService extends AbstractDiscoveryService {
+    private static final Logger logger = LoggerFactory.getLogger(HomematicDeviceDiscoveryService.class);
+    private static final int DISCOVER_TIMEOUT_SECONDS = 300;
+
+    private HomematicBridgeHandler bridgeHandler;
+    private Future<?> scanFuture;
+
+    public HomematicDeviceDiscoveryService(HomematicBridgeHandler bridgeHandler) {
+        super(ImmutableSet.of(new ThingTypeUID(BINDING_ID, "-")), DISCOVER_TIMEOUT_SECONDS, true);
+        this.bridgeHandler = bridgeHandler;
+    }
+
+    /**
+     * Called on component activation.
+     */
+    public void activate() {
+        super.activate(null);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void deactivate() {
+        super.deactivate();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void startScan() {
+        logger.debug("Starting Homematic discovery scan");
+        loadDevices();
+        waitForScanFinishing();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected synchronized void stopScan() {
+        logger.debug("Stopping Homematic discovery scan");
+        stopLoadDevices();
+        super.stopScan();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Starting Homematic discovery background scan");
+        loadDevices();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void stopBackgroundDiscovery() {
+        logger.debug("Stopping Homematic discovery background scan");
+        stopLoadDevices();
+    }
+
+    /**
+     * Stops the device scanning thread.
+     */
+    public void stopLoadDevices() {
+        if (scanFuture != null) {
+            scanFuture.cancel(true);
+            scanFuture = null;
+        }
+    }
+
+    /**
+     * Waits for the discovery scan to finish and then returns.
+     */
+    public void waitForScanFinishing() {
+        logger.debug("Waiting for finishing Homematic device discovery scan");
+        if (scanFuture != null) {
+            try {
+                scanFuture.get();
+            } catch (CancellationException ex) {
+                // ignore
+            } catch (Exception ex) {
+                logger.error("Error waiting for device discovery scan: {}", ex.getMessage(), ex);
+            }
+        }
+    }
+
+    /**
+     * Starts a thread which loads all Homematic devices connected to the gateway.
+     */
+    public void loadDevices() {
+        if (scanFuture == null) {
+            scanFuture = scheduler.submit(new Runnable() {
+
+                @Override
+                public void run() {
+                    try {
+                        removeOlderResults(getTimestampOfLastScan());
```

Comment:
wouldn't it be better to do that at the end, so that only devices are removed that were not discovered on this execution? If you do it at the start, all devices will disappear from the inbox only to reappear shortly after.